Update to 6.12 kernel

rbradford opened the pull request:
- **build: Update kernel to v6.12**
- **HACK: Force kernel rebuild**

rbradford committed:
build: Update kernel to v6.12

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

TimePrinciple commented:
Great work! 🫡
While it seems we don't have a `HACK` component yet

rbradford committed:
tests: Update interrupt group name on ARM

This has changed with kernel 6.12.

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford commented:
> Great work! 🫡 While it seems we don't have a `HACK` component yet

Good - I wouldn't want it to pass!

rbradford committed:
scripts: Temporarily build kernel as part of CI

Updating the kernel to v6.12 has shown up a flaw in the workflow for our
binary kernel releases. The CI job that builds the binary kernel in the
cloud-hypervisor/linux repository fetches the config from the main
branch of the cloud-hypervisor/cloud-hypervisor repository. However the
CI job to update the kernel version to use is in the cloud-hypervisor
repository.

As a workaround - update the kernel config and version in the
cloud-hypervisor repository to point to v6.12 and use the ability to
build the kernel during the CI run. Once merged to main a new release
can be made in the linux respository which will build a binary asset
using the new config. After that release the CI jobs on the
cloud-hypervisor repository can changed back to using the binary kernel
assets.

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

TimePrinciple commented:
> @likebreath @TimePrinciple See the details in the last commit about a flaw in the binary kernel workflow - this workaround is fine for now whilst I think about a better solution long term.

Noted, I'll try to propose a simpler workflow

rbradford commented:
> > @likebreath @TimePrinciple See the details in the last commit about a flaw in the binary kernel workflow - this workaround is fine for now whilst I think about a better solution long term.
> 
> Noted, I'll try to propose a simpler workflow - And again, thanks for this great work, really a life saver (I'm still struggling in backporting AIA stuffs to 6.6 kernel in other projects😂)

I'm thinking about embedding the version in the name of the config file here or storing the config in the fork as a new defconfig file.

TimePrinciple commented:
> > > @likebreath @TimePrinciple See the details in the last commit about a flaw in the binary kernel workflow - this workaround is fine for now whilst I think about a better solution long term.
> > 
> > 
> > Noted, I'll try to propose a simpler workflow - And again, thanks for this great work, really a life saver (I'm still struggling in backporting AIA stuffs to 6.6 kernel in other projects😂)
> 
> I'm thinking about embedding the version in the name of the config file here or storing the config in the fork as a new defconfig file.

IMHO if we keep the config files here with multiple version, that would in a way sending a signal we are maintaining all these kernel versions with current version of CLH, while we don't have CI setup for that purpose. If we were to keep old version configs here while leaving them not covered by our CI, maybe the best place for them is in the git history of CLH.

Thus I would prefer to move those configs to `cloud-hypervisor/linux` repo as you said and suggested by @likebreath, we finish kernel stuff there and focus on CLH here 😉

blitz commented:
> IMHO if we keep the config files here with multiple version, that would in a way sending a signal we are maintaining all these kernel versions with current version of CLH, while we don't have CI setup for that purpose. If we were to keep old version configs here while leaving them not covered by our CI, maybe the best place for them is in the git history of CLH.

Is there anything special about these Linux configs? We've been using distro kernels without any issue so far. If there is nothing speaking against distro kernels, that would be an easy source of pre-built kernels (that people will probably use anyway).

rbradford commented:
I bisected to the following commit between 6.6 to 6.7

```
commit d67e32f26713c35669e343edfdce6fa97a7d6bbc (HEAD, refs/bisect/bad)
Author: Mike Kravetz <mike.kravetz@oracle.com>
Date:   Wed Oct 18 19:31:04 2023 -0700

    hugetlb: restructure pool allocations
    
    Allocation of a hugetlb page for the hugetlb pool is done by the routine
    alloc_pool_huge_page.  This routine will allocate contiguous pages from a
    low level allocator, prep the pages for usage as a hugetlb page and then
    add the resulting hugetlb page to the pool.
    
    In the 'prep' stage, optional vmemmap optimization is done.  For
    performance reasons we want to perform vmemmap optimization on multiple
    hugetlb pages at once.  To do this, restructure the hugetlb pool
    allocation code such that vmemmap optimization can be isolated and later
    batched.
    
    The code to allocate hugetlb pages from bootmem was also modified to
    allow batching.
    
    No functional changes, only code restructure.
    
    Link: https://lkml.kernel.org/r/20231019023113.345257-3-mike.kravetz@oracle.com
    Signed-off-by: Mike Kravetz <mike.kravetz@oracle.com>
    Reviewed-by: Muchun Song <songmuchun@bytedance.com>
    Tested-by: Sergey Senozhatsky <senozhatsky@chromium.org>
    Cc: Anshuman Khandual <anshuman.khandual@arm.com>
    Cc: Barry Song <21cnbao@gmail.com>
    Cc: David Hildenbrand <david@redhat.com>
    Cc: David Rientjes <rientjes@google.com>
    Cc: James Houghton <jthoughton@google.com>
    Cc: Joao Martins <joao.m.martins@oracle.com>
    Cc: Konrad Dybcio <konradybcio@kernel.org>
    Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
    Cc: Miaohe Lin <linmiaohe@huawei.com>
    Cc: Michal Hocko <mhocko@suse.com>
    Cc: Naoya Horiguchi <naoya.horiguchi@linux.dev>
    Cc: Oscar Salvador <osalvador@suse.de>
    Cc: Usama Arif <usama.arif@bytedance.com>
    Cc: Xiongchun Duan <duanxiongchun@bytedance.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
```

Notice the "No functional changes, only code restructure." :-)

Need some more detailed investigation however I think it is related to huge pages and/or interaction with virtio-iommu itself.

TimePrinciple commented:
> I bisected to the following commit between 6.6 to 6.7
> 
> ```
> commit d67e32f26713c35669e343edfdce6fa97a7d6bbc (HEAD, refs/bisect/bad)
> Author: Mike Kravetz <mike.kravetz@oracle.com>
> Date:   Wed Oct 18 19:31:04 2023 -0700
> 
>     hugetlb: restructure pool allocations
>     
>     Allocation of a hugetlb page for the hugetlb pool is done by the routine
>     alloc_pool_huge_page.  This routine will allocate contiguous pages from a
>     low level allocator, prep the pages for usage as a hugetlb page and then
>     add the resulting hugetlb page to the pool.
>     
>     In the 'prep' stage, optional vmemmap optimization is done.  For
>     performance reasons we want to perform vmemmap optimization on multiple
>     hugetlb pages at once.  To do this, restructure the hugetlb pool
>     allocation code such that vmemmap optimization can be isolated and later
>     batched.
>     
>     The code to allocate hugetlb pages from bootmem was also modified to
>     allow batching.
>     
>     No functional changes, only code restructure.
>     
>     Link: https://lkml.kernel.org/r/20231019023113.345257-3-mike.kravetz@oracle.com
>     Signed-off-by: Mike Kravetz <mike.kravetz@oracle.com>
>     Reviewed-by: Muchun Song <songmuchun@bytedance.com>
>     Tested-by: Sergey Senozhatsky <senozhatsky@chromium.org>
>     Cc: Anshuman Khandual <anshuman.khandual@arm.com>
>     Cc: Barry Song <21cnbao@gmail.com>
>     Cc: David Hildenbrand <david@redhat.com>
>     Cc: David Rientjes <rientjes@google.com>
>     Cc: James Houghton <jthoughton@google.com>
>     Cc: Joao Martins <joao.m.martins@oracle.com>
>     Cc: Konrad Dybcio <konradybcio@kernel.org>
>     Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
>     Cc: Miaohe Lin <linmiaohe@huawei.com>
>     Cc: Michal Hocko <mhocko@suse.com>
>     Cc: Naoya Horiguchi <naoya.horiguchi@linux.dev>
>     Cc: Oscar Salvador <osalvador@suse.de>
>     Cc: Usama Arif <usama.arif@bytedance.com>
>     Cc: Xiongchun Duan <duanxiongchun@bytedance.com>
>     Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
> ```
> 
> Notice the "No functional changes, only code restructure." :-)
> 
> Need some more detailed investigation however I think it is related to huge pages and/or interaction with virtio-iommu itself.

I managed to get more log around where error occurs,
Terminal on the left is before the bisected commit, and the other one is on that commit:
<img width="1941" alt="image" src="https://github.com/user-attachments/assets/c55e13c1-4388-44f7-8cfe-a1e855c694ad" />
I'll need to fix broken `syscall_table` generation issue in `seccompiler` and I will get back to investigate this as soon as possible 🧐

rbradford committed:
virtio-devices: iommu: Search full range for GVA conversion

Remove an erroneous optimisation that used the page size mask to reduce
the range to iterate through on the set of mappings. This doesn't work
as the virtio-iommu ranges are larger than a single page. This may have
worked in the past when the mappings were limited to a single page.

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford committed:
virtio-devices: iommu: Use hex formatting in log messages

This means that the the addresses are more easily readable.

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford commented:
@likebreath @TimePrinciple Am happy to report that I figured out the issue with virtio-iommu and now the test_vfio test is passing so we can go ahead with this PR. I will address the issue of moving the kernel configuration files to the other repository and re-enabling the binary kernel use as a follow-up. FWIW this fix also allowed the use of booting from a block device behind a virtio-iommu - something that was broken even with v6.2

TimePrinciple commented:
Great work!

likebreath reviewed (approved):
Thank you for looking into this. @rbradford Please merge when you are ready.
